### PR TITLE
feat(one-login-auth-token): validates auth token claims and user sub

### PR DESCRIPTION
**Pull request by dylanwrightCO**
https://technologyprogramme.atlassian.net/browse/GAP-2069

- validates iss claim
- validates aud claim
- validates exp claim
- validates auth_time claim
- validates signature
- validates user sub from token matches user sub from userInfo

Tested against prod One Login environment

### CODE REVIEW

File: src/main/java/gov/cabinetofice/gapuserservice/service/OneLoginService.java [hunk 1]
```diff
@@ -282,6 +295,52 @@ public void migrateUser(final User user, final String jwt) {
...
+            String keyId = signedAuthToken.getHeader().getKeyID();
+
+            JWKSet jwkSet = JWKSet.load(new URL(oneLoginBaseUrl.concat("/.well-known/jwks.json")));
+            Optional<JWK> optionalMatchingJwk = Optional.ofNullable(jwkSet.getKeyByKeyId(keyId));
+            JWK matchingJwk = optionalMatchingJwk.orElseThrow(() -> new UnauthorizedException
+                    ("Matching JWK not found for key ID: " + keyId));
+
+            ECDSAVerifier verifier = new ECDSAVerifier((ECKey) matchingJwk);
+
+            if (!signedAuthToken.verify(verifier) || !signedAuthToken.getHeader().getAlgorithm().equals(jwtAlgorithm)) {
```
Comment (jgunnCO) [inline src/main/java/gov/cabinetofice/gapuserservice/service/OneLoginService.java:333]: invert these? no point trying to verify if the algorithm is wrong (and verification will always fail)

Comment (jgunnCO) [reply src/main/java/gov/cabinetofice/gapuserservice/service/OneLoginService.java:333]: oh actually is this ever going to fail?

```js
signedAuthToken.getHeader().getAlgorithm().equals(jwtAlgorithm)
```

we pull the algorithm out of the jwt header on line 323 so isn't this check is essentially circular?

Comment (dylanwrightCO) [reply src/main/java/gov/cabinetofice/gapuserservice/service/OneLoginService.java:333]: Yeah you're right, need to compare the Algorithm matches the JWKSet instead I think

Comment (jgunnCO) [reply src/main/java/gov/cabinetofice/gapuserservice/service/OneLoginService.java:333]: yeah

**Review by jgunnCO — DISMISSED** [review]

**Review by john-tco — APPROVED** [review]

**Review by john-tco — APPROVED** [review]

**Pull request merged**